File: two.py

def read_file(file):
    with open(file) as f:
        return [tuple(line.split()) for line in f]

class Sub():
    def __init__(self):
        self.pos = 0
        self.depth = 0
    def move(self, op):
        direction = op[0]
        distance = int(op[1])
        if direction == "forward":
            self.pos += distance
        elif direction == "down":
            self.depth += distance
        else:
            self.depth -= distance

class SubAim():
    def __init__(self):
        self.pos = 0
        self.depth = 0
        self.aim = 0
    def move(self, op):
        direction = op[0]
        amt = int(op[1])
        if direction == "forward":
            self.pos += amt
            self.depth += self.aim * amt
        elif direction == "down":
            self.aim += amt
        else:
            self.aim -= amt



def main():
    S, T = Sub(), SubAim()
    ops = read_file('./inputs/two.txt')
    for op in ops:
        S.move(op)
        T.move(op)
    print(f"Part 1: {S.pos * S.depth}")
    print(f"Part 2: {T.pos * T.depth}")

if __name__ == "__main__":
    main()

File: again-in-cpp-for-practice/two.cpp
#include <iostream>
#include <fstream>
#include <string>

using std::cout;
using std::string;

struct Sub
{
    int position, depth;
};

enum string_code
{
    cForward,
    cDown,
    cUp
};

string_code encode(string inString)
{
    if (inString == "forward")
        return cForward;
    if (inString == "down")
        return cDown;
    if (inString == "up")
        return cUp;
}

int main()
{
    std::ifstream infile("../inputs/two.txt");
    struct Sub sub;
    sub.position = 0;
    sub.depth = 0;
    string direction;
    int amount;
    while (infile >> direction >> amount)
    {
        switch (encode(direction))
        {
        case cForward:
            sub.position += amount;
            break;
        case cDown:
            sub.depth += amount;
            break;
        case cUp:
            sub.depth -= amount;
            break;
        default:
            break;
        }
    };
    cout << "Part 1 " << (sub.depth * sub.position) << std::endl;
    return 0;
    /* Skipping part 2 for now. This is a learning exercise for me and I won't learn any more C++ by doing part 2. Just add aim to sub and adjust movement */
};
